src/net/connection.rs

```diff
@@ -81,16 +127,117 @@ impl ActiveConnections {
+                );
+            });
     }
 
-    /// Returns true if the given connection exists.
-    pub fn exists(&self, address: &SocketAddr) -> bool {
-        self.connections.contains_key(&address)
+    pub fn update_connection_manager(
+        conn: &mut VirtualConnection,
+        sender: &Sender<ConnectionEvent<ReceiveEvent>>,
+        manager: &mut dyn SocketManager,
+        socket: &mut SocketWithConditioner,
+        time: Instant,
+        buffer: &mut [u8],
+    ) {
+        while let Some(changes) = conn.state_manager.update(buffer, time) {
+            match changes {
+                Ok(event) => match event {
+                    Either::Left(packet) => {
+                        // TODO properly handle error, instead of assert
+                        let packet = conn
+                            .process_outgoing(
+                                packet.packet_type,
+                                packet.payload,
+                                packet.delivery,
+                                packet.ordering,
+                                None,
+                                time,
+                            )
+                            .expect("connection manager packet should not fail");
+                        if let Outgoing::Packet(outgoing) = packet {
+                            socket.send_packet_and_log(
+                                &conn.remote_address,
+                                conn.state_manager.as_mut(),
+                                &outgoing.contents(),
+                                manager,
+                                "sending packet from connection manager",
+                            );
+                        } else {
+                            panic!("connection manager cannot send fragmented packets");
+                        }
+                    }
+                    Either::Right(state) => {
+                        if let Some(old) = conn.current_state.try_change(&state) {
+                            if let Err(err) = match &conn.current_state {
+                                ConnectionState::Connected(data) => sender.send(ConnectionEvent(
+                                    conn.remote_address,
+                                    ReceiveEvent::Connected(data.clone()),
+                                )),
+                                ConnectionState::Disconnected(closed_by) => {
+                                    sender.send(ConnectionEvent(
+                                        conn.remote_address,
+                                        ReceiveEvent::Disconnected(DisconnectReason::ClosedBy(
+                                            closed_by.clone(),
+                                        )),
+                                    ))
+                                }
+                                _ => panic!(
+                                    "Invalid state transition: {:?} -> {:?}",
```

Comment:
This is a tricky one.
It is a connection manager error, but it is created by laminar, so in order to properly handle it I need to provide some kind of severity levels to connection manager errors so that connection manager error handling could be independent of socket manager.

I came up with this modification:
```rust
/// Generic error type, that is used by ConnectionManager implementation.
#[derive(Debug, PartialEq, Clone)]
pub enum ConnectionManagerError {
    /// Something really bad has happened, this is not a recoverable error, and the connection should be destroyed.
    Fatal(String),
    /// Something unexpected has happened, but the connection is still in a valid state.
    /// `SocketManager` can decide when to destroy the connection if two many warnings are propagated from the same connection in a short amount of time.
    Warning(String)
    // TODO: is it enought?
}
```

Reply:
And changed this panic! to ConnectionManagerError::Fatal